gateway-provider-security-authc-remote/src/main/java/org/apache/knox/gateway/filter/RemoteAuthFilter.java

```diff
@@ -0,0 +1,249 @@
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletContext;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.security.KeyStore;
+import java.security.PrivilegedActionException;
+import java.security.PrivilegedExceptionAction;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+public class RemoteAuthFilter implements Filter {
+
+  private static final String CONFIG_REMOTE_AUTH_URL = "remote.auth.url";
+  private static final String CONFIG_INCLUDE_HEADERS = "remote.auth.include.headers";
+  private static final String CONFIG_CACHE_KEY_HEADER = "remote.auth.cache.key";
+  private static final String CONFIG_EXPIRE_AFTER = "remote.auth.expire.after";
+  private static final String DEFAULT_CACHE_KEY_HEADER = "Authorization";
+  private static final String CONFIG_USER_HEADER = "remote.auth.user.header";
+  private static final String CONFIG_GROUP_HEADER = "remote.auth.group.header";
+
+  private String remoteAuthUrl;
+  private List<String> includeHeaders;
+  private String cacheKeyHeader;
+  private String userHeader;
+  private String groupHeader;
+  /*
+  For Testing
+   */
+  HttpURLConnection httpURLConnection;
+
+  Cache<String, Subject> authenticationCache;
+
+  private static final AuditService auditService = AuditServiceFactory.getAuditService();
+  private static final Auditor auditor = auditService.getAuditor(
+          AuditConstants.DEFAULT_AUDITOR_NAME, AuditConstants.KNOX_SERVICE_NAME, AuditConstants.KNOX_COMPONENT_NAME );
+  private final RemoteAuthMessages LOGGER = MessagesFactory.get( RemoteAuthMessages.class );
+
+  @Override
+  public void init(FilterConfig filterConfig) throws ServletException {
+    remoteAuthUrl = filterConfig.getInitParameter(CONFIG_REMOTE_AUTH_URL);
+    if (remoteAuthUrl == null || remoteAuthUrl.isEmpty()) {
+      LOGGER.missingRequiredParameter(CONFIG_REMOTE_AUTH_URL);
+      throw new ServletException(CONFIG_REMOTE_AUTH_URL + " is a missing required param.");
+    }
+    includeHeaders = Arrays.asList(filterConfig.getInitParameter(CONFIG_INCLUDE_HEADERS).split(","));
+    cacheKeyHeader = filterConfig.getInitParameter(CONFIG_CACHE_KEY_HEADER) != null ? filterConfig
+            .getInitParameter(CONFIG_CACHE_KEY_HEADER) : DEFAULT_CACHE_KEY_HEADER;
+    String cachetime = filterConfig.getInitParameter(CONFIG_EXPIRE_AFTER);
+    if (cachetime != null) {
+      int expireAfterMinutes = Integer.parseInt(cachetime);
+      authenticationCache = CacheBuilder.newBuilder()
+              .expireAfterWrite(expireAfterMinutes, TimeUnit.MINUTES)
+              .build();
+    }
+
+    userHeader = filterConfig.getInitParameter(CONFIG_USER_HEADER);
+    if (userHeader == null || userHeader.isEmpty()) {
```

Comment:
Can this be a default value? e.g. `X-Knox-Actor-ID`

Reply:
Sure.